src/daemon/plugin.c

```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
 
-  double start = rf->rf_ctx.align_read; /* (default is -1.0) */
-
-  /* if AlignRead is not set, start reading immediately */
-  if (start < 0) {
+  /* If AlignRead is not set, start reading immediately. */
+  if (alignRead < 0) {
     rf->rf_next_read = cdtime();
-  } else {
-    cdtime_t now = cdtime();
-
-    /* get minute in hour and second in minute and millisecond in second */
-    struct timeval tv = CDTIME_T_TO_TIMEVAL(now);
-    uint64_t curr_msec =
-        (((tv.tv_sec / 60) % 60) * 60 // minute of hour in seconds
-         + gmtime(&(tv.tv_sec))->tm_sec) *
-            1000             // add seconds of minute and convert to ms
-        + tv.tv_usec / 1000; // add millisecond of second
-
-    /* DEBUG("%s: current time: %lu sec, %lu ms in minute, start time: %.3lf,
-    interval: %lu", rf->rf_name, tv.tv_sec, curr_msec, start,
-    CDTIME_T_TO_MS(rf->rf_interval)); */
-
-    /* determine delay until next (aligned) read */
-    int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(start) - MS_TO_CDTIME_T(curr_msec);
-    while (sleep_cdt < 0) {
-      sleep_cdt += rf->rf_interval;
+    return;
+  }
+
+  /* AlignRead values greater than Interval do not make sense. */
+  if (alignRead >= rf->rf_interval) {
+    static bool align_error = true;
+
+    if (align_error) {
+      ERROR("AlignRead has to be smaller than Interval (%.3lf >= %d)!",
+           alignRead, (int)rf->rf_interval);
+      align_error = false;
     }
 
-    /* set the next read time */
-    if (sleep_cdt > 0) {
-      rf->rf_next_read = now + sleep_cdt;
+    rf->rf_next_read = cdtime();
+    return;
+  }
 
-      DEBUG("%s plugin: wait %.3fs before first read at %ld.%03ds since epoche",
-            rf->rf_name, CDTIME_T_TO_DOUBLE(sleep_cdt),
-            (long int)(CDTIME_T_TO_TIMEVAL(rf->rf_next_read).tv_sec),
-            (int)(CDTIME_T_TO_TIMEVAL(rf->rf_next_read).tv_usec / 1000));
-    } else {
-      rf->rf_next_read = cdtime();
+  /* Seconds and minutes of Interval and AlignRead have to be perfect divisors
+   * of 60. Otherwise to perserve alignment. */
+  int sec_interval = CDTIME_T_TO_TIME_T(rf->rf_interval) % 60;
+  int sec_align = (long)alignRead % 60;
+  int min_interval = CDTIME_T_TO_TIME_T(rf->rf_interval) / 60;
+  int min_align = (int)(alignRead / 60);
+  if ((sec_interval > 0 && 60 % sec_interval != 0) ||
+      (sec_align > 0 && 60 % sec_align != 0) ||
+      (min_interval > 0 && 60 % min_interval != 0) ||
+      (min_align > 0 && 60 % min_align != 0)) {
+    static bool align_error = true;
+
+    if (align_error) {
+      ERROR("Seconds and minutes of AlignRead and Interval have to be perfect "
+            "divisors of 60 to perserve alignment.");
+      align_error = true;
     }
+
+    rf->rf_next_read = cdtime();
+    return;
   }
+
+  /* get minutes in current hour and seconds in current minute */
+  cdtime_t now = cdtime();
+  struct timeval now_tv = CDTIME_T_TO_TIMEVAL(now);
+  struct tm *now_tm;
+  now_tm = localtime(&now_tv.tv_sec);
```

Comment:
1) localtime() is not thread-safe.
2) look for  `CDTIME_T_TO_TIME_T`

Reply:
The code is executed before read threads are started. Nevertheless, I thought about using get_local_time() from utils_time.c, which would require more code changes.